### PR TITLE
Implement Discovered Hosts Entity

### PR DESCRIPTION
**Pull request by ldjebran**
related issue https://github.com/SatelliteQE/airgun/issues/150
test run output https://github.com/SatelliteQE/robottelo/pull/6252

### CODE REVIEW

File: airgun/views/discoveredhosts.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        not exist in the DOM. Also there is no welcome_message.
+        """
+        # This View is searchable if search field is displayed.
+        return self.searchbox.search_field.is_displayed
+
+
+class DiscoveredHostDetailsTableColumn(TableColumn):
+    """A Table column for Discovered host details Table Row"""
+    def __locator__(self):
+        return self.browser.element(
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply airgun/views/discoveredhosts.py]: There is no headers in this table in the conventional usage of this classes (please see https://github.com/SatelliteQE/airgun/pull/202/files#diff-858fb908f8feceba1dfa979beb727b09R89), there is a th in each row first column.

[the remark this answers is not in the document]
Comment (ldjebran) [reply airgun/views/discoveredhosts.py]: eg the rows contains the header and data

File: airgun/views/discoveredhosts.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    table = DiscoveredHostsViewTable(
+        './/table',
+        column_widgets={
+            0: Checkbox(
+                locator=".//input[contains(@id, 'host_ids')]"),
+            'Name': Text("./a"),
+        }
+    )
+    # This view has no welcome message
+    welcome_message = None
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply airgun/views/discoveredhosts.py]: Think this is not weird but logical, as we have no welcome message and we are inheriting that value.

[the remark this answers is not in the document]
Comment (oshtaier) [reply airgun/views/discoveredhosts.py]: 75% of other entities also doesn't, but we do not assign it to None or your case has something to do with custom `is_searchable` implementation?

[the remark this answers is not in the document]
Comment (ldjebran) [reply airgun/views/discoveredhosts.py]: yes in any case we do not want the usage of that message here.

File: airgun/views/discoveredhosts.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
+from wait_for import wait_for
+from widgetastic.widget import (
+    Checkbox,
+    Text,
+    Select,
+    Table,
+    TableRow,
+    TableColumn,
```
Comment (abalakh) [inline airgun/views/discoveredhosts.py]: please sort alphabetically

File: airgun/views/discoveredhosts.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+)
+from airgun.views.host import HostCreateView
+from airgun.widgets import (
+    ActionsDropdown,
+    FilteredDropdown,
+)
+
+
+class DiscoveredHostsViewTable(SatTable):
+    """Discovered hosts table that has a diffrent no rows message location.
```
Comment (abalakh) [inline airgun/views/discoveredhosts.py]: different

File: airgun/views/discoveredhosts.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        return False
+
+    @property
+    def headers(self):
+        """Explicitly returns the headers name and value"""
+        return 'name', 'value'
+
+    def read(self):
+        """Transform rows to a dict {property_name: property_value ...}."""
+        properties = super(DiscoveredHostDetailsTable, self).read()
```
Comment (abalakh) [inline airgun/views/discoveredhosts.py]: nitpick: there's shorter py3 version: `properties = super().read()`

File: airgun/entities/discoveredhosts.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        """Reboot discovered hosts which names supplied in entities_list."""
+        if not isinstance(entities_list, (list, tuple)):
+            entities_list = [entities_list]
+        view = self.navigate_to(
+            self, 'Reboot', entities_list=entities_list)
+        view.submit.click()
+        view.flash.assert_no_error()
+        view.flash.dismiss()
+
+    def assign_organization(self, entities_list, organization):
```
Comment (abalakh) [inline airgun/entities/discoveredhosts.py]: Just curious - why different from host's implementation of helpers/navigate steps/views, as the situation is kinda similar to me?
https://github.com/SatelliteQE/airgun/pull/179/files#diff-a4f1f5c4c861bb6e90ce95c2c5c915c7R71

Comment (ldjebran) [reply airgun/entities/discoveredhosts.py]: we implemented in hosts that way as host has too many actions, here in this PR we have a restraint number, if it we can make an arrangement that all entities action should be that way, no problem I'll update this PR, but we should take into consideration that all entities that has actions (except host) are implemented as this PR.

File: airgun/views/discoveredhosts.py (hunk 5 above)
Comment (ldjebran) [reply airgun/views/discoveredhosts.py]: we can go further more deeper, in py3 implantations, but prefer to select from which class we are using that read.

File: airgun/views/discoveredhosts.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+                for prop in properties if prop['value']}
+
+
+class DiscoveredHostDetailsView(BaseLoggedInView):
+    """Discovered Host details view"""
+    breadcrumb = BreadCrumb()
+    back = Text(
+        ".//a[contains(@class, 'btn') and @data-id='aid_discovered_hosts']")
+    actions = ActionsDropdown(
+        "//a[@data-toggle='dropdown']"
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply airgun/views/discoveredhosts.py]: yes it's not unique

**Comment by ldjebran** [comment]
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v tests/foreman/ui_airgun/test_discoveredhost.py 
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-3.6.1, py-1.5.4, pluggy-0.6.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: file
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.2.1, mock-1.6.3, forked-0.2, cov-2.5.1
collecting 8 items                                                                                           2018-09-18 14:58:02 - conftest - DEBUG - BZ deselect is disabled in settings

collected 8 items                                                                                            

tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_read_details_from_discovered_host PASSED [ 12%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_pxe_based_discovery PASSED               [ 25%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_pxe_less_with_dhcp_unattended PASSED     [ 37%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_provision_using_quick_host_button PASSED [ 50%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_update_name PASSED                       [ 62%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_manual_provision_host_with_rule PASSED   [ 75%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_delete PASSED                            [ 87%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_reboot PASSED                            [100%]

======================================== 8 passed in 1333.56 seconds =========================================
```
assign location and organization has been affected by different bugs in last snap

File: airgun/views/discoveredhosts.py [hunk 8]
```diff
@@ -232,8 +230,7 @@ class DiscoveredHostProvisionDialog(BaseLoggedInView):
 
     @property
     def is_displayed(self):
-        return self.browser.wait_for_element(
-            self.title, exception=False) is not None
+        return self.title.is_displayed
```
Comment (abalakh) [inline airgun/views/discoveredhosts.py]: What's the reason behind this change? It will probably lead to exceptions in logs when element is not found

Comment (ldjebran) [reply airgun/views/discoveredhosts.py]: It will not leed to exception and working as expected, the older implementation was not working always failing when calling is_displayed

Comment (ldjebran) [reply airgun/views/discoveredhosts.py]: now the tests are very robust in this area, and never fail any more. locally or on Saucelabs.

Comment (ldjebran) [reply airgun/views/discoveredhosts.py]: Sorry I confused with an other case, the problem here was that it navigated to all discovered hosts and the navigation stopped, as the older implementation was returning always True.

Comment (abalakh) [reply airgun/views/discoveredhosts.py]: We use `is_diplayed` just like that directly for `welcome_message` in searching, and i see pretty often messages like that in logs:
```
2018-09-27 17:52:20 - widgetastic_null - ERROR - [ContentViewVersionDetailsView/errata/welcome_message]: An exception happened during read call (elapsed 33 ms)
2018-09-27 17:52:20 - widgetastic_null - ERROR - [ContentViewVersionDetailsView/errata/welcome_message]: Message: Could not find an element Locator(by='xpath', locator="//div[@class='blank-slate-pf' or @id='welcome']")
Traceback (most recent call last):
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/browser.py", line 335, in element
    return elements[0]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/log.py", line 117, in wrapped
    result = f(self, *args, **kwargs)
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/widget.py", line 1064, in read
    return self.text
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/widget.py", line 1061, in text
    return self.browser.text(self, parent=self.parent)
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/browser.py", line 595, in text
    text = self.element(locator, *args, **kwargs).text
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/browser.py", line 327, in element
    elements = self.elements(locator, *args, **kwargs)
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/browser.py", line 818, in elements
    force_check_safe=force_check_safe)
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/browser.py", line 250, in elements
    locator = self._process_locator(locator)
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/browser.py", line 203, in _process_locator
    return locator.__element__()
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/widget.py", line 78, in wrapped
    return method(self, *new_args, **new_kwargs)
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/widget.py", line 353, in __element__
    return self.parent_browser.element(locator)
  File "/Users/andrii/workspace/widgetastic.core/src/widgetastic/browser.py", line 337, in element
    raise NoSuchElementException('Could not find an element {}'.format(repr(locator)))
selenium.common.exceptions.NoSuchElementException: Message: Could not find an element Locator(by='xpath', locator="//div[@class='blank-slate-pf' or @id='welcome']")
```

Comment (ldjebran) [reply airgun/views/discoveredhosts.py]: @abalakh it's just logged or the test fail at that line

**Comment by ldjebran** [comment]
@abalakh @oshtaier changes made
The last changes concern one comment about super user
the second about the is_displayed: The dialog is always displayed on the view in hidden sate but we need is_displayed for navigation success,

```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v tests/foreman/ui_airgun/test_discoveredhost.py 
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-3.6.1, py-1.5.4, pluggy-0.6.0 -- /home/dlezz/.pyenv/versions/3.6.6/envs/robottelo/bin/python
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: file
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.2.1, mock-1.10.0, forked-0.2, cov-2.5.1
collecting 8 items                                                                                           2018-10-01 15:19:16 - conftest - DEBUG - BZ deselect is disabled in settings

collected 8 items                                                                                            

tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_read_details_from_discovered_host PASSED [ 12%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_pxe_based_discovery PASSED               [ 25%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_pxe_less_with_dhcp_unattended PASSED     [ 37%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_provision_using_quick_host_button PASSED [ 50%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_update_name PASSED                       [ 62%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_manual_provision_host_with_rule PASSED   [ 75%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_delete PASSED                            [ 87%]
tests/foreman/ui_airgun/test_discoveredhost.py::test_positive_reboot PASSED                            [100%]

======================================== 8 passed in 1330.19 seconds =========================================
```
